### PR TITLE
feat: add dropdown to community ask

### LINKED ISSUE #31: Add course drop down to the new question page
When a user wants to ask a new question, beside the title of the question on the right hand side, ask the user to select a course (from the courses they belong to) and it should be a required field.

Once a course is associated with a question, make sure the query to fetch the questions are filtered by a course that the current user belongs to.

image

Requirements to work on this task:
- Frontend work
- Supabase related work.

Reach out to @rotimi-best if you are interested in picking this one

### PR DESCRIPTION
**Pull request by tunny17**
## What does this PR do?

Fixes #31 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [x] Enhancement (small improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

- Go to the community/ask route, select a drop down, ask the question and click on publish then go to community to see the questions but this time there's a small link to course of the respective question.

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [ ] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [ ] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the ClassroomIO Docs if changes were necessary
Screenshot (552)

### CODE REVIEW

File: apps/dashboard/src/lib/components/Org/Community/index.svelte [hunk 1]
```diff
@@ -9,11 +9,24 @@
...
 
   export let isLMS = false;
 
   let isLoading = false;
   let discussions = [];
+  let courseId = '';
+
+  // this function is to solve a situation where a student clicks the course title that he isn't enrolled in
+  function setPath(isLMS: boolean, discussion: any, courseId: string): string {
+    if (isLMS && discussion.course_id === courseId) {
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/components/Org/Community/index.svelte]: @tunny17 Please don't use snake case for `course_id`

File: apps/dashboard/src/lib/components/Org/Community/index.svelte [hunk 2]
```diff
@@ -23,20 +36,26 @@
...
         title,
         votes,
         created_at,
         slug,
         comments:community_answer(count),
         author:profile(
           fullname
+        ),
+        course_title:course(
+          title
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/components/Org/Community/index.svelte]: Change this to

```
course(title)
```

File: apps/dashboard/src/lib/components/Org/Community/index.svelte [hunk 3]
```diff
@@ -23,20 +36,26 @@
...
         )
       `
       )
       .eq('organization_id', orgId)
       .order('created_at', { ascending: false });
     console.log('data', data);
     console.log('error', error);
+    console.log();
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/components/Org/Community/index.svelte]: remove

File: apps/dashboard/src/lib/components/Org/Community/index.svelte [hunk 4]
```diff
@@ -47,12 +66,17 @@
     discussions =
       data?.map((discussion) => ({
         title: discussion.title,
+        course_id: discussion.course_id,
+        course_title: discussion.course_title.title,
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/components/Org/Community/index.svelte]: change these from snake to camel case

File: apps/dashboard/src/lib/components/Org/Community/index.svelte [hunk 5]
```diff
@@ -47,12 +66,17 @@
...
         slug: discussion.slug,
         author: discussion?.author?.fullname,
         comments: discussion.comments?.[0]?.count || 0,
         votes: discussion.votes,
         createdAt: calDateDiff(discussion.created_at)
       })) || [];
+
+    // sets course id of the current user
+    courseId = $courses[0].id;
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/components/Org/Community/index.svelte]: what is this used for?

File: apps/dashboard/src/lib/utils/services/org/index.ts [hunk 6]
```diff
@@ -200,6 +200,8 @@ export async function getCourseBySiteName(siteName: string) {
 }
 
 export async function getCurrentOrg(siteName: string, justGet = false) {
+  console.log(siteName);
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/services/org/index.ts]: remove

File: apps/dashboard/src/routes/+layout.server.ts [hunk 7]
```diff
@@ -37,6 +37,8 @@ export const load = async ({ url, cookies }): Promise<LoadOutput> => {
     response.isOrgSite = debugMode || answer;
     response.orgSiteName = debugMode ? _orgSiteName : subdomain;
     response.org = (await getCurrentOrg(response.orgSiteName, true)) || null;
+    console.log(response.org);
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/+layout.server.ts]: remove

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte [hunk 8]
```diff
@@ -10,15 +10,26 @@
...
     title?: string;
   } = {};
   let fields = {
     title: '',
-    body: ''
+    body: '',
+    course_id: ''
   };
 
+  let coursesResults;
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/lms/community/ask/+page.svelte]: what is the purpose of this variable?

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte [hunk 9]
```diff
@@ -49,6 +60,10 @@
       goto(`/lms/community/${slug}`);
     }
   }
+
+  onMount(() => {
+    getCourses($profile.id, $currentOrg.id);
+  });
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/lms/community/ask/+page.svelte]: onMount, it is possible that both arguments are not available in the store.

1. You should call them reactively
2. Make sure it is called when both arguments `$profile.id` and `$currentOrg.id` is not undefined

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte [hunk 10]
```diff
@@ -66,8 +81,24 @@
...
+      bind:value={fields.title}
+      placeholder="Title"
+      errorMessage={errors.title}
+      className="w-[75%]"
+    />
+    <select
+      class="w-[25%]"
+      bind:value={fields.course_id}
+      on:change={(e) => (fields.course_id = e.target.value)}
+    >
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/lms/community/ask/+page.svelte]: Use the select component from carbon components

File: apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte [hunk 11]
```diff
@@ -10,15 +10,29 @@
   import TextEditor from '$lib/components/TextEditor/index.svelte';
   import TextField from '$lib/components/Form/TextField.svelte';
   import { profile } from '$lib/utils/store/user';
+  import { fetchCourses } from '$lib/components/Courses/api';
+  import { onMount } from 'svelte';
+
+  // export let isLMS = false;
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: remove

File: apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte [hunk 12]
```diff
@@ -49,6 +63,10 @@
       goto(`${$currentOrgPath}/community/${slug}`);
     }
   }
+
+  onMount(() => {
+    getCourses($profile.id, $currentOrg.id);
+  });
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: Follow the review I gave here: apps/dashboard/src/routes/lms/community/ask/+page.svelte

File: apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte [hunk 13]
```diff
@@ -69,8 +87,24 @@
...
+      placeholder="Title"
+      errorMessage={errors.title}
+      className="w-[75%]"
+    />
+    <select
+      class="w-[25%]"
+      bind:value={fields.course_id}
+      on:change={(e) => (fields.course_id = e.target.value)}
+    >
+      {#if fetchedCourses}
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: this if statement is not necessary

**Review by rotimi-best — CHANGES_REQUESTED** [review]
Left you some reviews. Please pay attention to them and let me know if you have any questions.

File: apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte (hunk 11 above)
Comment (tunjiadeyemi) [reply apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: i can't remove the fetchCourse, remember that's where i'm getting all the course titles to display on the dropdown

Comment (rotimi-best) [reply apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: This comment has nothing to do with fetchCourses but the commented code

Comment (tunjiadeyemi) [reply apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: oh, yeah, i removed it already, you highlighted line 13 - 16 so i had to say something there.

File: apps/dashboard/src/lib/components/Org/Community/index.svelte [hunk 14]
```diff
@@ -55,9 +75,32 @@
...
+  />
+  <Dropdown
+    class="w-[25%]"
+    size="xl"
+    label="Select Course"
+    items={allCourses.map((course) => ({ id: course.id, text: course.title }))}
+    bind:selectedId
+    on:select={(event) => {
+      const { selectedId } = event.detail;
+    }}
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/components/Org/Community/index.svelte]: @tunny17 is this needed?

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte [hunk 15]
```diff
@@ -66,8 +87,25 @@
...
+    <Dropdown
+      class="w-[25%]"
+      size="xl"
+      label="Select Course"
+      invalid={isInvalid}
+      items={fetchedCourses.map((course) => ({ id: course.id, text: course.title }))}
+      bind:selectedId={fields.course_id}
+      on:select={(event) => {
+        const { selectedId } = event.detail;
+        fields.course_id = selectedId;
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/lms/community/ask/+page.svelte]: @tunny17 What is this logic for?

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte (hunk 8 above)
Comment (rotimi-best) [reply apps/dashboard/src/routes/lms/community/ask/+page.svelte]: @tunny17 Waiting for your response on this

**Review by rotimi-best — COMMENTED** [review]
Great progress.

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte [hunk 16]
```diff
@@ -49,6 +64,12 @@
       goto(`/lms/community/${slug}`);
     }
   }
+
+  $: {
+    if ($profile.id !== undefined && $currentOrg.id !== undefined) {
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/lms/community/ask/+page.svelte]: You can shorten this to:

```js
if ($profile.id && $currentOrg.id) {
```

File: apps/dashboard/src/lib/components/Org/Community/index.svelte (hunk 14 above)
Comment (tunjiadeyemi) [reply apps/dashboard/src/lib/components/Org/Community/index.svelte]: no, i've fixed it.

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte (hunk 15 above)
Comment (tunjiadeyemi) [reply apps/dashboard/src/routes/lms/community/ask/+page.svelte]: i've deleted it, it's gone.

Comment (tunjiadeyemi) [reply apps/dashboard/src/routes/lms/community/ask/+page.svelte]: @rotimi-best  i'm not sure why i can't reply the comment directly below but that variable is set to hold the data coming from the fetchCourses request:
"
  async function getCourses(userId: string | null, orgId: string) {
    coursesResults = await fetchCourses(userId, orgId);
    fetchedCourses = coursesResults.allCourses;
  }
"
(line 30 to 33).
note: the request is to display all courses in the select dropdown

Comment (rotimi-best) [reply apps/dashboard/src/routes/lms/community/ask/+page.svelte]: Yes I know why you call the fetchCourses function, however you don't use `coursesResults` anywhere in the code which is why I can't understand why the variable is there

File: apps/dashboard/src/lib/utils/functions/validator.js [hunk 17]
```diff
@@ -17,6 +17,10 @@ const askCommunityValidationSchema = z.object({
   body: z.string().min(20, {
     message: 'Must be 10 or more characters long',
     invalid_type_error: 'Must not be empty'
+  }),
+  course_id: z.string().min(36, {
+    message: 'Invalid course',
+    invalid_type_error: 'Select a course'
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/functions/validator.js]: btw your validation will fail because the field is `courseId` not `course_id`

File: apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte [hunk 18]
```diff
@@ -10,23 +11,39 @@
...
+  }
+
   async function handleSave() {
     errors = askCommunityValidation(fields);
     console.log('handleSave errors', errors);
+    isInvalid = true;
 
     if (Object.keys(errors).length) {
       return;
     }
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: `isInvalid` should be inside of the if statement and not before you verify that there was an actual error

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte (hunk 15 above)
Comment (tunjiadeyemi) [reply apps/dashboard/src/routes/lms/community/ask/+page.svelte]: the courseResults only stores the data from fetchCourse and fetchedCourses gets the allCourses property from the courseResults object

File: apps/dashboard/src/lib/utils/functions/validator.js (hunk 17 above)
Comment (tunjiadeyemi) [reply apps/dashboard/src/lib/utils/functions/validator.js]: alright, noted.

File: apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte (hunk 18 above)
Comment (tunjiadeyemi) [reply apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: alright, noted.

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte (hunk 15 above)
Comment (rotimi-best) [reply apps/dashboard/src/routes/lms/community/ask/+page.svelte]: I see that, but for you to create a top variable outside the function scope, it has to be used to render the UI or used in another function. I asking to make you realize that it isn't needed.

Please change it to:

```js
async function getCourses(userId: string | null, orgId: string) {
  const fetchCoursesResponse = await fetchCourses(userId, orgId);
  fetchedCourses = fetchCoursesResponse.allCourses;
}
```

Then remove `coursesResults` from the file and also do that in any other place this implemented

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte [hunk 19]
```diff
@@ -10,23 +11,37 @@
...
 
   let errors: {
     title?: string;
+    courseId?: string;
   } = {};
   let fields = {
     title: '',
-    body: ''
+    body: '',
+    course_id: ''
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/lms/community/ask/+page.svelte]: @tunny17 No, don't change this to snake case because of the validation, you should have just changed the validation to `courseId`, instead of the fields and the UI code to snake case

File: apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte [hunk 20]
```diff
@@ -10,23 +11,39 @@
...
 
   let errors: {
     title?: string;
+    course_id?: string;
   } = {};
   let fields = {
     title: '',
-    body: ''
+    body: '',
+    course_id: ''
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: Same thing here with the `course_id`, change to `courseId`

**Review by rotimi-best — CHANGES_REQUESTED** [review]

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte [hunk 21]
```diff
@@ -66,8 +87,21 @@
...
+      bind:value={fields.title}
+      placeholder="Title"
+      errorMessage={errors.title}
+      className="w-[75%]"
+    />
+    <Dropdown
+      class="w-[25%]"
+      size="xl"
+      label="Select Course"
+      invalid={isInvalid}
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/lms/community/ask/+page.svelte]: btw, thinking about it now, you don't need this `isInvalid` variable.  You should just do `!!errors.courseId`. This value already holds the error state of the field, there is no need to create another variable for error state.

I am assuming you have changed the variable in the validation to `courseId` instead of `course_id`.

File: apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte [hunk 22]
```diff
@@ -69,8 +92,21 @@
...
+      bind:value={fields.title}
+      placeholder="Title"
+      errorMessage={errors.title}
+      className="w-[75%]"
+    />
+    <Dropdown
+      class="w-[25%]"
+      size="xl"
+      label="Select Course"
+      invalid={isInvalid}
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/org/[slug]/community/ask/+page.svelte]: Same thing here about `isInvalid`

**Review by rotimi-best — CHANGES_REQUESTED** [review]

**Comment by rotimi-best** [comment]
@tunny17 I also noticed that when testing the community question and I click on Edit, I can't change the course id.

Please add the course dropdown on edit state as well of a community question

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte (hunk 15 above)
Comment (tunjiadeyemi) [reply apps/dashboard/src/routes/lms/community/ask/+page.svelte]: oh, okay, noted

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte (hunk 19 above)
Comment (tunjiadeyemi) [reply apps/dashboard/src/routes/lms/community/ask/+page.svelte]: i had to change to fields.course_id from fields.courseId becasue the name of the column on on the supabase table is "course_id", it was a bug but i can rename the column on the table to courseId if  you want me to

File: apps/dashboard/src/routes/lms/community/[slug]/+page.svelte [hunk 23]
```diff
@@ -227,7 +241,11 @@
 
   async function handleQuestionEdit() {
     if (isEditMode) {
-      const fields = { title: editContent.title, body: editContent.body };
+      const fields = {
+        title: editContent.title,
+        body: editContent.body,
+        course_id: editContent.course_id
```
Comment (rotimi-best) [inline apps/dashboard/src/routes/lms/community/[slug]/+page.svelte]: @tunny17 I keep telling you to change this to camel case, the 3rd time.

bro what's up??? I am just wasting my time everytime on these PR reviews.

File: apps/dashboard/src/lib/utils/functions/validator.js (hunk 17 above)
Comment (rotimi-best) [reply apps/dashboard/src/lib/utils/functions/validator.js]: You haven't changed this???

File: apps/dashboard/src/routes/lms/community/ask/+page.svelte (hunk 19 above)
Comment (rotimi-best) [reply apps/dashboard/src/routes/lms/community/ask/+page.svelte]: no keep the column that way, but map it back to camel case when you get it from the backend

**Review by rotimi-best — CHANGES_REQUESTED** [review]
Please deal with the `course_id` issue. That's the last bug here

**Review by rotimi-best — APPROVED** [review]
LGTM
